### DISCUSSION TITLE
What's the best way to toggle buffer + grid view with vertico-multiform?

### DISCUSSION THREAD

**Discussion by deejayem** [minad/vertico#528] [Q&A] [answered]
I just discovered vertico-multiform, which is great (thank you!).

One thing I noticed is that while it's possible to combine buffer and grid display, e.g. with `(setq vertico-multiform-categories  '((file buffer grid)))`, I couldn't see an easy way to toggle this on and off, like there is with buffer and grid individually.

For example, it's not possible to do something like https://github.com/minad/vertico/blob/main/extensions/vertico-multiform.el#L204-L218 because there's no vertico-buffer-grid-mode (or similar).

This these seem to work (i.e. I can get from vertical display to buffer + grid display, and back again, by hitting the same key binding twice):
```
  (defun vertico-multiform-buffer-grid ()
    (interactive)
    (if (equal '(vertico-buffer-mode vertico-grid-mode) (car vertico-multiform--stack))
        (vertico-multiform-vertical)
      (push '(vertico-buffer-mode vertico-grid-mode) vertico-multiform--stack)
      (vertico-multiform--toggle 1)))
```

I thought this was also working, but it's not (i.e. switching back doesn't work):
```
  (defun vertico-multiform-buffer-grid ()
    (interactive)
    (if (equal '(vertico-buffer-mode vertico-grid-mode) (car vertico-multiform--stack))
        (progn (pop vertico-multiform--stack)
               (vertico-multiform--toggle 2))
      (push '(vertico-buffer-mode vertico-grid-mode) vertico-multiform--stack)
      (vertico-multiform--toggle 1)))
```

Are either of these a sensible approach, or is there a better way?

**Comment by minad** [comment]
Thanks.

> Are either of these a sensible approach, or is there a better way?

If it works it sounds as if it is. ;)

**Reply by deejayem** [reply]
One thing I noticed after creating the discussion, is that after using the first version, if I re-run `(vertico-multiform-mode +1)` I get:
`Warning (emacs): vertico-multiform state is inconsistent`

Which suggests that something isn't right. I'm not sure if it matters though, since everything seems to work correctly.

**Reply by minad** [reply]
Indeed then your command should probably be improved to preserve the consistency of the state. In case you manage to fix it, feel free to add the command to the wiki.

**Reply by deejayem** [reply, marked as answer]
It looks like the problem was that I misunderstood how `vertico-multiform--stack` is being used, and I needed to use `setcar` instead of `push`. The following seems to do a better job, so I'll add it to the wiki later:

```
  (defun vertico-multiform-buffer-grid ()
    "Toggle displaying Vertico as a grid in a large window (like a regular buffer)."
    (interactive)
    (if (equal '(vertico-buffer-mode vertico-grid-mode) (car vertico-multiform--stack))
        (vertico-multiform-vertical)
      (setcar vertico-multiform--stack '(vertico-buffer-mode vertico-grid-mode))
      (vertico-multiform--toggle 1)))
```
